src/compressed_chunk.c

```diff
@@ -33,6 +35,117 @@ void Compressed_FreeChunk(Chunk_t *chunk) {
     free(chunk);
 }
 
+static void swapChunks(CompressedChunk *a, CompressedChunk *b) {
+    CompressedChunk tmp = *a;
+    *a = *b;
+    *b = tmp;
+}
+
+static void extendChunk(ChunkResult res, CompressedChunk *chunk, Sample *sample) {
+    if (res != CR_OK) {
+        int oldsize = chunk->size;
+        chunk->size += 64;
+        chunk->data = (u_int64_t *)realloc(chunk->data, chunk->size * sizeof(char));
+        memset((char *)chunk->data + oldsize, 0, 64);
+        printf("Chunk extended to %lu \n", chunk->size);
+        ChunkResult result = Compressed_AddSample(chunk, sample);
+        assert(result == CR_OK);
+    }
+}
+
+static void trimChunk(CompressedChunk *chunk, int minSize) {
+    int excess = chunk->size - (chunk->idx) / BIT;
+
+    assert(excess >= 0); // else we have written beyond allocated memory
+
+    if (excess > 0) {
+        size_t newSize = max(chunk->size - excess + 2, minSize);
+        chunk->data = realloc(chunk->data, newSize);
+        chunk->size = newSize;
+    }
+}
+
+Chunk_t *Compressed_SplitChunk(Chunk_t *chunk) {
+    CompressedChunk *curChunk = chunk;
+    size_t split = curChunk->count / 2;
+    size_t curNumSamples = curChunk->count - split;
+
+    // add samples in new chunks
+    size_t i = 0;
+    Sample sample;
+    ChunkIter_t *iter = Compressed_NewChunkIterator(curChunk, false);
+    CompressedChunk *newChunk1 = Compressed_NewChunk(curChunk->size / sizeof(Sample));
+    CompressedChunk *newChunk2 = Compressed_NewChunk(curChunk->size / sizeof(Sample));
+    for (; i < curNumSamples; ++i) {
+        Compressed_ChunkIteratorGetNext(iter, &sample);
+        ChunkResult res = Compressed_AddSample(newChunk1, &sample);
+        extendChunk(res, newChunk1, &sample);
+    }
+    for (; i < curChunk->count; ++i) {
+        Compressed_ChunkIteratorGetNext(iter, &sample);
+        ChunkResult res = Compressed_AddSample(newChunk2, &sample);
+        extendChunk(res, newChunk2, &sample);
+    }
+
+    trimChunk(newChunk1, 0);
+    trimChunk(newChunk2, 0);
+    swapChunks(curChunk, newChunk1);
+
+    Compressed_FreeChunkIterator(iter, false);
+    Compressed_FreeChunk(newChunk1);
+
+    return newChunk2;
+}
+
+ChunkResult Compressed_UpsertSample(UpsertCtx *uCtx, int *size) {
+    *size = 0;
+    ChunkResult rv = CR_OK;
+    ChunkResult addRes = CR_OK;
+    ChunkResult nextRes = CR_OK;
+    CompressedChunk *oldChunk = (CompressedChunk *)uCtx->inChunk;
+
+    size_t newSize = oldChunk->size / sizeof(Sample) + 64;
```

Comment:
@ashtul lets remove the +64 now that we fixes the issue with the realloc.

Reply:
OMG
sorry